File: FilFeed/src/main/java/com/phicdy/filfeed/ui/FeedUrlHookActivity.java
package com.phicdy.filfeed.ui;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.widget.Toast;

import com.phicdy.filfeed.R;
import com.phicdy.filfeed.db.DatabaseAdapter;
import com.phicdy.filfeed.rss.Feed;
import com.phicdy.filfeed.task.NetworkTaskManager;

public class FeedUrlHookActivity extends Activity {

	private ProgressDialog dialog;
	private BroadcastReceiver receiver;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_feed_url_hook);

		Intent intent = getIntent();
		if (intent != null && intent.getAction().equals(Intent.ACTION_VIEW)) {
			String url = intent.getDataString();
			if (url != null) {
				dialog = new ProgressDialog(this);
				dialog.setProgressStyle(ProgressDialog.STYLE_SPINNER);
				dialog.setMessage(getString(R.string.adding_feed));

				receiver = new BroadcastReceiver() {

					@Override
					public void onReceive(Context context, Intent intent) {
						String action = intent.getAction();
						if (action.equals(NetworkTaskManager.FINISH_ADD_FEED)) {
							DatabaseAdapter dbAdapter = DatabaseAdapter.getInstance(getApplicationContext());
							Feed newFeed = dbAdapter.getFeedByUrl(intent.getStringExtra(NetworkTaskManager.ADDED_FEED_URL));
							if (newFeed == null) {
								Toast.makeText(getApplicationContext(),
										R.string.add_feed_error,
										Toast.LENGTH_SHORT).show();
							} else {
								Toast.makeText(getApplicationContext(),
										R.string.add_feed_success,
										Toast.LENGTH_SHORT).show();
							}
							dialog.dismiss();
							finish();
						}
					}
				};
				IntentFilter filter = new IntentFilter();
				filter.addAction(NetworkTaskManager.FINISH_ADD_FEED);
				registerReceiver(receiver, filter);

				dialog.show();
				NetworkTaskManager.getInstance(getApplicationContext()).addNewFeed(url);
			}
		}
	}

	@Override
	protected void onPause() {
		super.onPause();
		if (receiver != null) {
			unregisterReceiver(receiver);
		}
	}

}

File: feature_feed_url_hook/src/main/java/com/phicdy/mycuration/feedurlhook/FeedUrlHookActivity.kt
package com.phicdy.mycuration.feedurlhook

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.annotation.StringRes
import androidx.annotation.UiThread
import androidx.appcompat.app.AppCompatActivity
import com.phicdy.mycuration.domain.rss.RssUrlHookIntentData
import com.phicdy.mycuration.feature.util.changeTheme
import com.phicdy.mycuration.tracker.TrackerHelper
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.AndroidEntryPoint
import dagger.hilt.android.components.ActivityComponent
import dagger.hilt.android.qualifiers.ActivityContext
import dagger.hilt.android.scopes.ActivityScoped
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import javax.inject.Inject
import kotlin.coroutines.CoroutineContext

@AndroidEntryPoint
class FeedUrlHookActivity : AppCompatActivity(), FeedUrlHookView,
    CoroutineScope {

    private val job = Job()
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Main

    @Inject
    lateinit var presenter: FeedUrlHookPresenter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_feed_url_hook)
        launch { presenter.create() }
    }

    override fun onResume() {
        super.onResume()
        changeTheme()
    }

    override fun showInvalidUrlErrorToast() {
        showToastOnUiThread(R.string.add_rss_error_invalid_url, Toast.LENGTH_SHORT)
    }

    override fun showGenericErrorToast() {
        showToastOnUiThread(R.string.add_rss_error_generic, Toast.LENGTH_SHORT)
    }

    override fun finishView() {
        finish()
    }

    override fun trackFailedUrl(url: String) {
        TrackerHelper.sendFailedParseUrl(getString(R.string.add_rss_from_intent_error), url)
    }

    override fun showSuccessToast() {
        showToastOnUiThread(R.string.add_rss_success, Toast.LENGTH_SHORT)
    }

    @UiThread
    private fun showToastOnUiThread(@StringRes res: Int, toastLength: Int) {
        runOnUiThread { Toast.makeText(applicationContext, res, toastLength).show() }
    }

    @Module
    @InstallIn(ActivityComponent::class)
    object FeedUrlHookModule {

        @ActivityScoped
        @Provides
        fun provideFeedUrlHookView(@ActivityContext activity: Context): FeedUrlHookView =
            activity as FeedUrlHookView

        @ActivityScoped
        @Provides
        fun provideRssUrlHookIntentData(@ActivityContext activity: Context): RssUrlHookIntentData =
                RssUrlHookIntentData(
                        action = (activity as AppCompatActivity).intent.action ?: "",
                        dataString = activity.intent.dataString ?: "",
                        extrasText = activity.intent.extras?.getCharSequence(Intent.EXTRA_TEXT, "")
                                ?: ""
                )
    }
}
